source/common/http/http2/codec_impl.cc

```diff
@@ -900,51 +905,78 @@ void ConnectionImpl::sendPendingFrames() {
-
-  std::vector<nghttp2_settings_entry> iv;
-
-  if (http2_settings.allow_connect_) {
-    iv.push_back({NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL, 1});
-  }
-
-  if (http2_settings.hpack_table_size_ != NGHTTP2_DEFAULT_HEADER_TABLE_SIZE) {
-    iv.push_back({NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, http2_settings.hpack_table_size_});
-    ENVOY_CONN_LOG(debug, "setting HPACK table size to {}", connection_,
-                   http2_settings.hpack_table_size_);
-  }
-
-  if (http2_settings.max_concurrent_streams_ != NGHTTP2_INITIAL_MAX_CONCURRENT_STREAMS) {
-    iv.push_back({NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_settings.max_concurrent_streams_});
-    ENVOY_CONN_LOG(debug, "setting max concurrent streams to {}", connection_,
-                   http2_settings.max_concurrent_streams_);
-  }
-
-  if (http2_settings.initial_stream_window_size_ != NGHTTP2_INITIAL_WINDOW_SIZE) {
-    iv.push_back(
-        {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_settings.initial_stream_window_size_});
-    ENVOY_CONN_LOG(debug, "setting stream-level initial window size to {}", connection_,
-                   http2_settings.initial_stream_window_size_);
-  }
+void ConnectionImpl::sendSettings(
+    const envoy::config::core::v3::Http2ProtocolOptions& http2_options, bool disable_push) {
+  std::unordered_set<nghttp2_settings_entry, SettingsEntryHash, SettingsEntryEquals> settings;
 
+  // Universally disable receiving push promise frames as we don't currently support
+  // them. nghttp2 will fail the connection if the other side still sends them.
+  // TODO(mattklein123): Remove this when we correctly proxy push promise.
+  // NOTE: This is a special case with respect to custom parameter overrides in that server push is
+  // not supported and therefore not end user configurable.
   if (disable_push) {
-    // Universally disable receiving push promise frames as we don't currently support them. nghttp2
-    // will fail the connection if the other side still sends them.
-    // TODO(mattklein123): Remove this when we correctly proxy push promise.
-    iv.push_back({NGHTTP2_SETTINGS_ENABLE_PUSH, 0});
+    settings.insert({static_cast<int32_t>(NGHTTP2_SETTINGS_ENABLE_PUSH), disable_push ? 0U : 1U});
+  }
+
+  for (const auto it : http2_options.custom_settings_parameters()) {
+    ASSERT(it.identifier().value() <= std::numeric_limits<uint16_t>::max());
+    if (it.identifier().value() == NGHTTP2_SETTINGS_ENABLE_PUSH && it.value().value() == 1) {
+      // For simplicity, warn when server push is configured enabled on either client or server
+      // connections. It _is_ actually enabled on downstream connections although Envoy itself will
+      // never send push_promise frames, but the nuance could be confusing to end users.
+      ENVOY_CONN_LOG(
+          warn,
+          "server push is not supported by Envoy and can not be enabled via a SETTINGS parameter.",
+          connection_);
+      continue;
```

Comment:
Could we just detect/verify this on config load and then fail the config instead of doing this here? Same below. It seems like all of the duplication logic could also be verified at config load time on either the upstream cluster side or the HCM side?

Reply:
Yeah, this is being done. I've simplified the validation logic.